app/controllers/responses_controller.rb

```diff
@@ -268,11 +349,61 @@ def odk_awaiting_media?
     params["*isIncomplete*"] == "yes"
   end
 
+  # Renders an Enketo form instead of a NEMO form.
+  def enketo
+    # Fail fast if we're on the wrong node version (this happens most often in development).
+    raise "Error: Unexpected Node version #{`node -v`}" unless `node -v`.match?("v16")
+
+    # Enketo can't render anything if we haven't rendered it to XML (e.g. unpublished draft).
+    if @response.form.odk_xml.blank?
+      flash[:error] = t("activerecord.errors.models.response.no_form_xml")
+      return redirect_to(params.permit!.merge("enketo": ""))
+    end
+
+    # This check is here until we have a way to encode legacy editor responses as ODK XML.
+    if action_name == "edit" && !@response.odk_xml.attached?
+      flash[:error] = t("activerecord.errors.models.response.no_response_xml")
+      return redirect_to(params.permit!.merge("enketo": ""))
+    end
+
+    @enketo_form_obj = enketo_form_obj
+    @enketo_instance_str = enketo_instance_str
+    @read_only = read_only?
+
+    # Fail fast if something went wrong with the CLI process.
+    raise RuntimeError unless @enketo_form_obj.present?
+
+    render(:enketo_form)
+  end
+
+  # The blank form template.
+  # Returns a string that's safe to print in a JS script.
+  def enketo_form_obj
+    # Terrapin seems to return an ASCII-encoded string, so we must interpret it
+    # as UTF-8 in order for the rest of the page to work for some kinds of forms.
+    command = Terrapin::CommandLine.new("node", ":transformer :xml")
```

Comment:
are there any security risks here in running something in the command line?

Reply:
or does Terrapin handle risks?

Reply:
yes, terrapin's tagline is: "Run shell commands safely, even with user-supplied values". it's published by thoughtbot and seems quite trusted.